feature: ngx.errlog: added the raw_log() API

thibaultcha opened the pull request:
This adds the `errlog.rawlog()` API function to allow for custom logging facilities, as discussed in openresty/lua-nginx-module#1074.

Its signature is:

```lua
errlog.rawlog(lvl, msg)
```

The second argument `msg` must be a string. This is a deliberate choice vs providing a variadic function like `ngx.log(lvl, ...)`,  to defer all formatting decisions (separator, string coertion, `__tostring()` call, ...) to the user when developing his or her logging facility.

Let me know your thoughts, and when everything will be agreed upon I will provide the documentation for this API.

agentzh commented:
@thibaultcha Where is the sister PR for the lua-nginx-module changes?

thibaultcha commented:
Just awaiting approval for the design of this API first, as any major change to its signature would imply changes to the underlying function as well...

agentzh commented:
@thibaultcha Yeah, this PR alone looks good to me.

thibaultcha commented:
Ok, sister PR just proposed!

thibaultcha commented:
@agentzh I have added some documentation for this function

agentzh commented:
@thibaultcha Great! I'll have a look when I have a chance. Thanks! Currently very busy with the company projects.

thibaultcha commented:
Yes ;)

thibaultcha commented:
Hi @agentzh!

Congrats on the latest 1.11.2.5 release. Sorry for the intrusion, I know a lot has been happening on OpenResty's side recently, and with this milestone reached, I was hoping you all had less on your plates and some time for a review on the following PRs:

- This one, and its sister PR openresty/lua-nginx-module#1106
- openresty/lua-cjson#26 for the `array_mt` feature
- openresty/lua-cjson#28 for the decoding of JSON arrays as `array_mt` Lua-land tables
- openresty/openresty#247 for the Nginx core resolver hosts file patch

Thanks!

agentzh commented:
@thibaultcha Thanks for the reminder! We're currently busy opensourcing the shiny new `ngx_stream_lua` module (and `ngx_meta_lua` module) sponsored by Mashape, Inc and upgrading OpenResty's nginx core to 1.13.4+. We'll get to your pending PRs as soon as we can manage.

thibaultcha commented:
Yep, I was aware :) Just an innocent reminder - would love to send more! Thanks for the ack!

thibaultcha commented on lib/ngx/errlog.lua:
```diff
@@ -104,4 +108,27 @@ function _M.get_sys_filter_level()
+    local rc = C.ngx_http_lua_ffi_raw_log(r, level, msg, #msg, err, errlen)
+
+    if rc == FFI_ERROR then
+        error("bad log level: " .. level, 2)
```
This should be:
```lua
error(err, 2)
```

thibaultcha replied:
Done (with `ffi.string` instead, for obvious reasons)

thibaultcha commented:
@agentzh I just rebased this PR against `master`, and rebased its sister PR against ngx_lua's `master` as well. Travis-CI is passing for this PR but the GitHub UI seems to not update (when I opened this PR it failed due to an ephemeral failure in the test suite iirc):

- https://travis-ci.org/openresty/lua-resty-core/builds/305479502
- sister PR seems to have failed due to an unrelated ephemeral failure: https://travis-ci.org/openresty/lua-nginx-module/builds/305478755

Would you be able to take another look at this PR soon? I believe it to be one more step towards opening new doors for OpenResty-based applications :) Thanks!

thibaultcha commented:
@agentzh Managed to fix the Travis status on the PR UI. Let me know

agentzh commented:
@thibaultcha Thanks for the update! It's been in my review queue.

agentzh commented on lib/ngx/errlog.md:
```diff
@@ -253,6 +254,108 @@ end
 
 [Back to TOC](#table-of-contents)
 
+rawlog
```
We use snake case function names consistently in this module. I think it's better to be `raw_error_log`?

agentzh replied:
Or `raw_log()`.

thibaultcha replied:
`raw_log()` makes perfect sense to me. I will update!

thibaultcha commented:
@agentzh Renamed the API to `raw_log()`. Should we also add a test asserting that `get_log()` can capture logs from this new API as well?

agentzh commented:
@thibaultcha Yeah, that's a good idea!

thibaultcha commented:
@agentzh I have just added such a test to the new suite for `raw_log()`.

thibaultcha commented:
@agentzh Ping if this is still of interest :) (very excited to use this in various libraries and projects!)

agentzh commented:
@thibaultcha Merged by tweaking the code as per @p0pr0ck5's feedback. Thanks!
